cub/cub/util_device.cuh

```diff
@@ -69,6 +70,172 @@ CUB_NAMESPACE_BEGIN
+   * passed to the agent, specialized for the case when we can use native shared memory as temporary
+   * storage.
+   */
+  static __device__ __forceinline__ typename agent_t::TempStorage&
+  get_temp_storage(typename agent_t::TempStorage& static_temp_storage, vsmem_t&)
+  {
+    return static_temp_storage;
+  }
+
+  /**
+   * @brief Used from within the device algorithm's kernel to get the temporary storage that can be
+   * passed to the agent, specialized for the case when we have to use global memory-backed
+   * virtual shared memory as temporary storage.
+   */
+  static __device__ __forceinline__ typename agent_t::TempStorage&
+  get_temp_storage(cub::NullType& static_temp_storage, vsmem_t& vsmem)
+  {
+    return *reinterpret_cast<typename agent_t::TempStorage*>(
+      static_cast<char*>(vsmem.gmem_ptr) + (vsmem_per_block * blockIdx.x));
+  }
+
+  /**
+   * @brief Hints to discard modified cache lines of the used virtual shared memory.
+   * modified cache lines. 
+   *
+   * @note Needs to be followed by `__syncthreads()` if the function returns true and the virtual shared memory is
+   * supposed to be reused after this function call.
+   */
+  template <bool needs_vsmem_ = needs_vsmem, typename ::cuda::std::enable_if<!needs_vsmem_, int>::type = 0>
+  static __device__ __forceinline__ bool discard_temp_storage(typename agent_t::TempStorage& temp_storage)
+  {
+    return false;
+  }
+
+  /**
+   * @brief Hints to discard modified cache lines of the used virtual shared memory.
+   * modified cache lines.
+   *
+   * @note Needs to be followed by `__syncthreads()` if the function returns true and the virtual shared memory is
+   * supposed to be reused after this function call.
+   */
+  template <bool needs_vsmem_ = needs_vsmem, typename ::cuda::std::enable_if<needs_vsmem_, int>::type = 0>
+  static __device__ __forceinline__ bool discard_temp_storage(typename agent_t::TempStorage& temp_storage)
+  {
+    // Ensure all threads finished using temporary storage
+    __syncthreads();
+
+    constexpr std::size_t line_size = 128;
+    const std::size_t linear_tid    = threadIdx.x;
+    const std::size_t block_stride  = line_size * blockDim.x;
+
+    char* ptr = reinterpret_cast<char*>(&temp_storage);
+
+    const auto size = static_cast<std::size_t>(sizeof(typename agent_t::TempStorage));
+    auto ptr_end    = ptr + size;
+    ptr_end -= (reinterpret_cast<std::uintptr_t>(ptr_end) % line_size);
+
+    std::size_t misalignment   = reinterpret_cast<std::uintptr_t>(ptr) % line_size;
+    std::size_t aligned_offset = (misalignment == 0) ? 0ULL : (line_size - misalignment);
+    ptr += aligned_offset;
+
+    // 128 byte-aligned virtual shared memory discard
+    for (auto thread_ptr = ptr + (linear_tid * line_size); thread_ptr < ptr_end; thread_ptr += block_stride)
+    {
+      cuda::discard_memory(thread_ptr, line_size);
+    }
+
+    return true;
```

Comment:
suggestion: we could bump vsmem temp storage size to multiple of 128 and avoid math here. This might also help with computing vsmem offset for packed user types.

Reply:
That's a great idea. Virtual shared memory is now padded to 128 bytes 👍